storage/innobase/row/row0ins.cc

```diff
@@ -2704,37 +2704,35 @@ row_ins_clust_index_entry_low(
 	DBUG_EXECUTE_IF("row_ins_row_level", goto skip_bulk_insert;);
 
 	if (!(flags & BTR_NO_UNDO_LOG_FLAG)
-	    && page_is_empty(block->page.frame)
+            && page_is_empty(block->page.frame)
+	    && block->page.id().page_no() == index->page
 	    && !entry->is_metadata() && !trx->duplicates
 	    && !trx->check_unique_secondary && !trx->check_foreigns
 	    && !trx->dict_operation
-	    && block->page.id().page_no() == index->page
-	    && !index->table->skip_alter_undo
-	    && !index->table->n_rec_locks
-	    && !index->table->is_active_ddl()
-	    && !index->table->has_spatial_index()
-	    && !index->table->versioned()
-            && (!dict_table_is_partition(index->table)
+	    && !index->table->is_native_online_ddl()
+	    && (!dict_table_is_partition(index->table)
 	        || thd_sql_command(trx->mysql_thd) == SQLCOM_INSERT)) {
-		DEBUG_SYNC_C("empty_root_page_insert");
 
-		trx->bulk_insert = true;
+		if (!index->table->has_spatial_index()
+		    && !index->table->n_rec_locks
+		    && !index->table->versioned()) {
 
-		if (!index->table->is_temporary()) {
+			ut_ad(!index->table->skip_alter_undo);
+			trx->bulk_insert= true;
+			if (index->table->is_temporary()) {
+				goto skip_bulk_insert;
```

Comment:
It would be good to add a comment to explain why we assign `trx->bulk_insert = true` for temporary tables. By the way, this function is formatted in the archaic InnoDB style, so there should be a space before the `=`.

It would also be less confusing if we renamed the `skip_bulk_insert` label into `row_level_insert`.

If there were no need to assign `trx->bulk_insert = true` for temporary tables, then we should check `!is_temporary()` before `!has_spatial_index()`. I see that the assignment was there also prior to this change.